grafana/grafana
# ExtSvcAuth: Clean up orphaned external services on start up

**What is this feature?**

This PR adds a new 1-time background job that cleans up orphaned external services on start up.

Orphaned external services can exist when their associated plugins have been removed through the CLI:
```
 ./grafana/bin/grafana-cli plugins remove plugin-to-remove
```

**Which issue(s) does this PR fix?**:

Fixes #https://github.com/grafana/grafana-enterprise/issues/5989

**Special notes for your reviewer:**

How to test this:
- clone the https://github.com/grafana/grafana-plugin-examples repository
- build and install the `app-with-service-account` and `app-with-on-behalf-of-auth` plugins:
  ```bash
  cd grafana-plugin-examples/examples/app-with-service-account/
  yarn build && mage -v
  cd ../app-with-on-behalf-of-auth
  yarn build && mage -v
  ln -sfn PATH_TO/grafana-plugin-examples/examples/app-with-on-behalf-of-auth/dist/ PATH_TO/grafana/data/plugins/grafana-appwithonbehalfofauth-app
  ln -sfn PATH_TO/grafana-plugin-examples/examples/app-with-service-account/dist/ PATH_TO/grafana/data/plugins/grafana-appwithserviceaccount-app
   ```
- Edit the config ini file with feature flags:
  ```ini
  [feature_toggles]
  enable = externalServiceAccounts externalServiceAuth
  ```
- Run Grafana. Navigate to `plugins` and check both plugins are there (you can enable them). Navigate to `serviceaccounts` and check that both plugins' service accounts are there.
- Remove one plugin:
  ```bash
  ./grafana/bin/grafana-cli plugins remove grafana-appwithserviceaccount-app
  # Removing plugin: grafana-appwithserviceaccount-app
  # Please restart Grafana after installing or removing plugins. Refer to Grafana documentation for instructions if necessary.
  ```
- Restart grafana and check that the service account for `grafana-appwithserviceaccount-app` has been deleted.
- You should also see the following logs:
  ```log
  logger=extsvcauth.registry t=2023-11-14T17:02:35.20268547+01:00 level=info msg="Detected removed External Service" service=grafana-appwithserviceaccount-app provider=ServiceAccounts
  logger=serviceauth.extsvcaccounts t=2023-11-14T17:02:35.202924547+01:00 level=info msg="Delete service account" service=grafana-appwithserviceaccount-app orgID=1 saID=16
  logger=serviceauth.extsvcaccounts t=2023-11-14T17:02:35.27970295+01:00 level=info msg="Deleted external service account" service=grafana-appwithserviceaccount-app orgID=1
  ```
- Remove the other plugin:
  ```bash
  ./grafana/bin/grafana-cli plugins remove grafana-appwithonbehalfofauth-app
  # Removing plugin: grafana-appwithonbehalfofauth-app
  # Please restart Grafana after installing or removing plugins. Refer to Grafana documentation for instructions if necessary.
  ```
- Restart grafana and check that the service account for `grafana-appwithserviceaccount-app` has been deleted.
- You should also see the following logs:
  ```log
  logger=extsvcauth.registry t=2023-11-14T16:54:41.544612593+01:00 level=info msg="Detected removed External Service" service=grafana-appwithonbehalfofauth-app provider=OAuth2Server
  logger=oauthserver t=2023-11-14T16:54:41.544699244+01:00 level=info msg="Remove external service" service=grafana-appwithonbehalfofauth-app
  logger=serviceauth.extsvcaccounts t=2023-11-14T16:54:42.554682225+01:00 level=info msg="Delete service account" service=grafana-appwithonbehalfofauth-app orgID=1 saID=17
  logger=serviceauth.extsvcaccounts t=2023-11-14T16:54:42.565611239+01:00 level=info msg="Deleted external service account" service=grafana-appwithonbehalfofauth-app orgID=1
  ```
  

Please check that:
